### PR TITLE
Mqtt5 Subscribe Operation

### PR DESCRIPTION
**Pull request by xiazhvera**
*Issue #, if available:*

*Description of changes:*
- Add subscribe operation
- Update MqttClientPublishRecievedEvents

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 1]
```diff
@@ -236,6 +240,55 @@ public class PublishPacket: CStruct {
...
+                identifier?.append(subscription_identifier)
+            }
+
+            let userProperties = convertOptionalUserProperties(
+                count: publishView.user_property_count,
+                userPropertiesPointer: publishView.user_properties)
+
+            guard let qos = QoS(rawValue: Int(publishView.qos.rawValue)) else {
+                fatalError("PublishPacket Received has an invalid qos")
+            }
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:271]: Can we use nativeValue here? Having both is confusing.

**Review by waahm7 — COMMENTED** [review]
I haven't reviewed everything yet.

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 2]
```diff
@@ -349,6 +484,34 @@ public class SubackPacket {
         self.reasonString = reasonString
         self.userProperties = userProperties
     }
+
+    static func convertFromNative(_ from: UnsafePointer<aws_mqtt5_packet_suback_view>?) -> SubackPacket? {
```
[the remark this answers is not in the document]
Comment (waahm7) [reply Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:488]: Same for all `convertFromNative` function.

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Client.swift [hunk 3]
```diff
@@ -58,6 +58,47 @@ public class Mqtt5Client {
...
+            }
+
+            subscribePacket.withCPointer { subscribePacketPointer in
+                var callbackOptions = aws_mqtt5_subscribe_completion_options()
+                callbackOptions.completion_callback = subscribeCompletionCallback
+                callbackOptions.completion_user_data = ContinuationCore(continuation: continuation).passRetained()
+                let result = aws_mqtt5_client_subscribe(rawValue, subscribePacketPointer, &callbackOptions)
+                if result != 0 {
+                    continuation.resume(throwing: CommonRunTimeError.crtError(CRTError(code: -1)))
+                }
```
[the remark this answers is not in the document]
Comment (xiazhvera) [reply Source/AwsCommonRuntimeKit/mqtt/Mqtt5Client.swift:96]: What's the benefit using guard instead of if?

[the remark this answers is not in the document]
Comment (waahm7) [reply Source/AwsCommonRuntimeKit/mqtt/Mqtt5Client.swift:96]: It gives a compile time guarantee of not falling through in the case of error. Continuation.resume is not a returning statement but we want to make sure that we don't do anything after resuming the continuation.

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (xiazhvera) [reply Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:488]: Thining about, it is easier to handle the nil case with the static function instead of the init function. With the init function, we need either explicitly check the nil value or handle exception for the nil case.

[the remark this answers is not in the document]
Comment (waahm7) [reply Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:488]: It should be the same. You will need to init?

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 4]
```diff
@@ -220,14 +220,18 @@ public class PublishPacket: CStruct {
...
                                 raw_publish_view.content_type = cContentType
-                                raw_publish_view.correlation_data = cCorrelationData
                                 raw_publish_view.response_topic = cResponseTopic
 
+                                if let _correlationData = self.correlationData {
+                                    return _correlationData.withAWSByteCursorPointer { cCorrelationData in
+                                        raw_publish_view.correlation_data = UnsafePointer(cCorrelationData)
+                                        return body(raw_publish_view)
+                                    }
+                                }
```
[the remark this answers is not in the document]
Comment (xiazhvera) [reply Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift]: Added a `withOptionalByteCursorPointerFromData` as correlation_data is a `Data` type.

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Client.swift [hunk 5]
```diff
@@ -65,6 +65,49 @@ public class Mqtt5Client {
...
+                continuationCore.continuation.resume(returning: suback)
+
+            }
+
+            subscribePacket.withCPointer { subscribePacketPointer in
+                var callbackOptions = aws_mqtt5_subscribe_completion_options()
+                callbackOptions.completion_callback = subscribeCompletionCallback
+                callbackOptions.completion_user_data = ContinuationCore(continuation: continuation).passRetained()
+                let result = aws_mqtt5_client_subscribe(rawValue, subscribePacketPointer, &callbackOptions)
+                guard result == 0 else {
```
Comment (sbSteveK) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Client.swift:103]: 0 -> AWS_OP_SUCCESS

**Review by sbSteveK — APPROVED** [review]

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Client.swift [hunk 6]
```diff
@@ -65,6 +65,49 @@ public class Mqtt5Client {
...
+
+            }
+
+            subscribePacket.withCPointer { subscribePacketPointer in
+                var callbackOptions = aws_mqtt5_subscribe_completion_options()
+                callbackOptions.completion_callback = subscribeCompletionCallback
+                callbackOptions.completion_user_data = ContinuationCore(continuation: continuation).passRetained()
+                let result = aws_mqtt5_client_subscribe(rawValue, subscribePacketPointer, &callbackOptions)
+                guard result == AWS_OP_SUCCESS else {
+                    return continuation.resume(throwing: CommonRunTimeError.crtError(CRTError.makeFromLastError()))
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Client.swift:104]: Memory leak: Need to release the ContinuationCore here.

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 7]
```diff
@@ -220,22 +220,74 @@ public class PublishPacket: CStruct {
...
+                            contentType) { cResponseTopic, cContentType in
                                 raw_publish_view.content_type = cContentType
-                                raw_publish_view.correlation_data = cCorrelationData
                                 raw_publish_view.response_topic = cResponseTopic
 
-                                return body(raw_publish_view)
+                                return withOptionalByteCursorPointerFromData(to: self.correlationData) {  cCorrelationData in
+                                    if let _cCorrelationData = cCorrelationData {
+                                        raw_publish_view.correlation_data = _cCorrelationData
+                                    }
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:232]: You don't need the if. We will just assign `raw_publish_view.correlation_data` to nil if cCorrelationData is nil.

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 8]
```diff
@@ -287,11 +339,67 @@ public class Subscription {
...
+
+    private var topicFilterBuffer = aws_byte_buf()
+
+    typealias RawType = aws_mqtt5_subscription_view
+    func withCStruct<Result>(_ body: (RawType) -> Result) -> Result {
+        var view = aws_mqtt5_subscription_view()
+        view.qos = self.qos.nativeValue
+        view.no_local = self.noLocal ?? false
+        view.retain_as_published = self.retainAsPublished ?? false
+        if let _retainType = self.retainHandlingType {
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:357]: What does _variableName naming scheme represent? Maybe just get rid of _?
```suggestion
        if let retainType = self.retainHandlingType {
```

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 9]
```diff
@@ -287,11 +339,67 @@ public class Subscription {
...
+            view.retain_handling_type = _retainType.natvieValue
+        } else {
+            view.retain_handling_type = aws_mqtt5_retain_handling_type(0)
+        }
+
+        return withByteCursorFromStrings(self.topicFilter) { _ in
+            view.topic_filter = aws_byte_cursor_from_buf(&topicFilterBuffer)
+            return body(view)
+        }
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:366]: Why call the with* function if you don't need the result?

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 10]
```diff
@@ -328,6 +436,32 @@ public class SubscribePacket {
                   subscriptionIdentifier: subscriptionIdentifier,
                   userProperties: userProperties)
     }
+
+    typealias RawType = aws_mqtt5_packet_subscribe_view
+    func withCStruct<Result>(_ body: (RawType) -> Result) -> Result {
+        var raw_subscrbe_view = aws_mqtt5_packet_subscribe_view()
+        return self.subscriptions.withCSubscriptions { subscriptionPointer in
+            raw_subscrbe_view.subscriptions =
+            UnsafePointer<aws_mqtt5_subscription_view>(subscriptionPointer)
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:445]: nit: bad formatting.

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 11]
```diff
@@ -328,6 +436,32 @@ public class SubscribePacket {
...
+            raw_subscrbe_view.subscription_count = self.subscriptions.count
+
+            return withOptionalUserPropertyArray(
+                of: userProperties) { userPropertyPointer in
+
+                    if let _userPropertyPointer = userPropertyPointer {
+                        raw_subscrbe_view.user_property_count = userProperties!.count
+                        raw_subscrbe_view.user_properties =
+                        UnsafePointer<aws_mqtt5_user_property>(_userPropertyPointer)
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:454]: nit: bad format.

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 12]
```diff
@@ -328,6 +436,32 @@ public class SubscribePacket {
...
+        var raw_subscrbe_view = aws_mqtt5_packet_subscribe_view()
+        return self.subscriptions.withCSubscriptions { subscriptionPointer in
+            raw_subscrbe_view.subscriptions =
+            UnsafePointer<aws_mqtt5_subscription_view>(subscriptionPointer)
+            raw_subscrbe_view.subscription_count = self.subscriptions.count
+
+            return withOptionalUserPropertyArray(
+                of: userProperties) { userPropertyPointer in
+
+                    if let _userPropertyPointer = userPropertyPointer {
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:451]: If it works
```suggestion
                    if let userPropertyPointer = userPropertyPointer {
```

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 13]
```diff
@@ -349,6 +483,34 @@ public class SubackPacket {
         self.reasonString = reasonString
         self.userProperties = userProperties
     }
+
+    public static func convertFromNative(_ from: UnsafePointer<aws_mqtt5_packet_suback_view>?) -> SubackPacket? {
+        if let _from = from {
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:488]: A guard will be better here instead of nesting 20 lines of code plus again, makes more sense to make it a failable init?.

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 14]
```diff
@@ -349,6 +483,34 @@ public class SubackPacket {
...
+            let subackPointer = _from.pointee
+
+            var subackReasonCodes: [SubackReasonCode] = []
+            for i in 0..<subackPointer.reason_code_count {
+                let reasonCodePointer = subackPointer.reason_codes.advanced(by: Int(i)).pointee
+                guard let reasonCode = SubackReasonCode(rawValue: Int(reasonCodePointer.rawValue))
+                else {fatalError("SubackPacket from native has an invalid reason code.")}
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:495]: nit: bad format. Is there a way to enforce/check this?

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 15]
```diff
@@ -220,22 +220,74 @@ public class PublishPacket: CStruct {
...
                             }
                         }
                     }
                 }
             }
         }
     }
+
+    static func convertFromNative(_ from: UnsafePointer<aws_mqtt5_packet_publish_view>?) -> PublishPacket? {
+        if let _from = from {
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:244]: Same comments as the other one. 
1. Guard will be better
2. If you make this a failable init?, you won't need to make different variables to pass to the init function, you can assign it directly like s`elf.publishView = `.

File: Test/AwsCommonRuntimeKitTests/mqtt/Mqtt5ClientTests.swift [hunk 16]
```diff
@@ -47,6 +47,7 @@ class Mqtt5ClientTests: XCBaseTestCase {
         public var onLifecycleEventConnectionFailure: OnLifecycleEventConnectionFailure?
         public var onLifecycleEventDisconnection: OnLifecycleEventDisconnection?
 
+        public let semaphorePublishReceived:DispatchSemaphore
```
Comment (waahm7) [inline Test/AwsCommonRuntimeKitTests/mqtt/Mqtt5ClientTests.swift:50]: nit: bad format. Can we add more rules to our format checker?

File: Source/AwsCommonRuntimeKit/crt/Utilities.swift [hunk 17]
```diff
@@ -90,6 +90,16 @@ func withOptionalAWSByteCursorFromData<Result>(
     }
 }
 
+func withOptionalByteCursorPointerFromData<Result>(
+    to data: Data?, _ body: (UnsafePointer<aws_byte_cursor>?) throws -> Result) rethrows -> Result {
+    guard let _data = data else {
+        return try body(nil)
+    }
```
[the remark this answers is not in the document]
Comment (xiazhvera) [reply Source/AwsCommonRuntimeKit/crt/Utilities.swift:97]: Debatable: I feel more safe to have a different name for a different variable type to avoid confusion. If it is not a strong against, I'd prefer keep it like this.

[the remark this answers is not in the document]
Comment (waahm7) [reply Source/AwsCommonRuntimeKit/crt/Utilities.swift:97]: Plus, I think it's better to shadow because you are not going to use data after the unwrap, so it should be unwrapped throughout the function, and it is not ideal to have both an optional and an unwrapped value in the scope if only one is needed and used.

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Client.swift [hunk 18]
```diff
@@ -65,6 +65,49 @@ public class Mqtt5Client {
...
 
+    /// Tells the client to attempt to subscribe to one or more topic filters.
+    ///
+    /// - Parameters:
+    ///     - subscribePacket: SUBSCRIBE packet to send to the server
+    /// - Returns:
+    ///     - @return Suback packet if the subscription operation succeed otherwise errorCode
+    ///
+    /// - Throws: CommonRuntimeError.crtError
+    public func subscribe(subscribePacket: SubscribePacket) async throws -> SubackPacket {
```
[the remark this answers is not in the document]
Comment (xiazhvera) [reply Source/AwsCommonRuntimeKit/mqtt/Mqtt5Client.swift:76]: I will leave it for now, as we plan to remove the parameter variable name in a future PR: 
`public func subscribe(_ packet: SubscribePacket) async throws -> SubackPacket {`

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift (hunk 15 above)
Comment (xiazhvera) [reply Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift]: Failable init is a good idea. As we have a list of PRs chained after this one, such big update would affect the following PRs. I will have a separate PR to make all of the "convertFromNative" to failable init.

File: Source/AwsCommonRuntimeKit/crt/Utilities.swift [hunk 19]
```diff
@@ -81,15 +81,25 @@ extension Data {
 
 func withOptionalAWSByteCursorFromData<Result>(
     to data: Data?, _ body: (aws_byte_cursor) throws -> Result) rethrows -> Result {
-    guard let _data = data else {
+    guard let data = data else {
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/crt/Utilities.swift:84]: nit: same for others if it's the same name
```suggestion
    guard let data else {
```

Comment (waahm7) [inline Source/AwsCommonRuntimeKit/crt/Utilities.swift:84]: The function name doesn't represent what it does because it will always return a byte cursor. Maybe rename it to 
`withAWSByteCursorFromOptionalData`

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 20]
```diff
@@ -148,7 +148,7 @@ public class PublishPacket: CStruct {
                 messageExpiryInterval: TimeInterval? = nil,
                 topicAlias: UInt16? = nil,
                 responseTopic: String? = nil,
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:150]: Not part of this PR but we should double check that we are not making anything a String if it can be non-utf8.

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 21]
```diff
@@ -179,63 +179,105 @@ public class PublishPacket: CStruct {
...
+        raw_publish_view.qos = self.qos.nativeValue
         raw_publish_view.retain = retain
         return topic.withByteCursor { topicCustor in
             raw_publish_view.topic =  topicCustor
             return withOptionalAWSByteCursorFromData(to: payload) { cByteCursor in
                 raw_publish_view.payload = cByteCursor
 
                 let _payloadFormatIndicatorInt: aws_mqtt5_payload_format_indicator? =
-                                                                    payloadFormatIndicator?.rawValue ?? nil
+                                                                    payloadFormatIndicator?.nativeValue ?? nil
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:190]: You don't need the ??. `?.` will return nill if it's nil. See https://docs.swift.org/swift-book/documentation/the-swift-programming-language/optionalchaining/

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 22]
```diff
@@ -179,63 +179,105 @@ public class PublishPacket: CStruct {
...
-
+                    raw_publish_view.payload_format = payloadPointer
+                    raw_publish_view.message_expiry_interval_seconds = messageExpiryIntervalPointer
+                    raw_publish_view.topic_alias = topicAliasPointer
+                        
                     return withOptionalArrayRawPointer(of: subscriptionIdentifiers) { subscriptionPointer in
 
-                        if let _subscriptionPointer = subscriptionPointer {
-                            raw_publish_view.subscription_identifiers = _subscriptionPointer
+                        if let subscriptionPointer = subscriptionPointer {
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:203]: Better to just do `if let subscriptionPointer, subscriptionCount {
}` instead of force unwrap.

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 23]
```diff
@@ -179,63 +179,105 @@ public class PublishPacket: CStruct {
...
+                            raw_publish_view.subscription_identifiers = subscriptionPointer
                             raw_publish_view.subscription_identifier_count = subscriptionIdentifiers!.count
                         }
 
                         return withOptionalUserPropertyArray(
                             of: userProperties) { userPropertyPointer in
-                            if let _userPropertyPointer = userPropertyPointer {
+                            if let userPropertyPointer = userPropertyPointer {
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:210]: Same as above

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 24]
```diff
@@ -179,63 +179,105 @@ public class PublishPacket: CStruct {
...
                                 raw_publish_view.user_properties =
-                                    UnsafePointer<aws_mqtt5_user_property>(_userPropertyPointer)
+                                    UnsafePointer<aws_mqtt5_user_property>(userPropertyPointer)
                             }
-                            return withOptionalByteCursorPointerFromString(
+                            return withOptionalByteCursorPointerFromStrings(
                             responseTopic,
-                            correlationData,
-                            contentType) { cResponseTopic, cCorrelationData, cContentType in
+                            contentType) { cResponseTopic, cContentType in
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:217]: nit weird format
responseTopic should not align with return and should have at least one tab.

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 25]
```diff
@@ -179,63 +179,105 @@ public class PublishPacket: CStruct {
...
                     }
                 }
             }
         }
     }
+
+    static func convertFromNative(_ from: UnsafePointer<aws_mqtt5_packet_publish_view>?) -> PublishPacket? {
+        guard let from = from else {
+            return nil
+        }
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:236]: if it works
```suggestion
        guard let from  else {
            return nil
        }
```

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 26]
```diff
@@ -179,63 +179,105 @@ public class PublishPacket: CStruct {
...
+        let publishView = from.pointee
+
+        let payload: Data = Data(bytes: publishView.payload.ptr, count: publishView.payload.len)
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:239]: nit / debatable: If the type is really obvious, I think it makes it cleaner to just omit it or at least keep it consistent.
Same comment for others.
```suggestion
        let payload = Data(bytes: publishView.payload.ptr, count: publishView.payload.len)
```

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 27]
```diff
@@ -179,63 +179,105 @@ public class PublishPacket: CStruct {
...
+
+        let payloadFormatIndicator: PayloadFormatIndicator? = publishView.payload_format != nil ?
+        PayloadFormatIndicator(rawValue: Int(publishView.payload_format.pointee.rawValue)) : nil
+
+        let messageExpiryInterval = convertOptionalUInt32(publishView.message_expiry_interval_seconds)
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:244]: Not relevant to this PR, but I would prefer that we get rid of all the convert* (either removing them or refactoring them) by the final merge. We should avoid introducing new terminologies. For example, if we need a function like this, it should be an extension over Optional, where Element == UInt32 instead of a separate function.

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 28]
```diff
@@ -328,6 +424,32 @@ public class SubscribePacket {
...
+        var raw_subscrbe_view = aws_mqtt5_packet_subscribe_view()
+        return self.subscriptions.withCSubscriptions { subscriptionPointer in
+            raw_subscrbe_view.subscriptions =
+                UnsafePointer<aws_mqtt5_subscription_view>(subscriptionPointer)
+            raw_subscrbe_view.subscription_count = self.subscriptions.count
+
+            return withOptionalUserPropertyArray(
+                of: userProperties) { userPropertyPointer in
+
+                    if let userPropertyPointer = userPropertyPointer {
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:439]:
```suggestion
                    if let userPropertyPointer, userProperties {
```

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 29]
```diff
@@ -349,6 +471,36 @@ public class SubackPacket {
         self.reasonString = reasonString
         self.userProperties = userProperties
     }
+
+    public static func convertFromNative(_ from: UnsafePointer<aws_mqtt5_packet_suback_view>?) -> SubackPacket? {
+        
+        guard let from = from else {
+            return nil
+        }
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:479]:
```suggestion
        guard let from else {
            return nil
        }
```

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 30]
```diff
@@ -435,17 +587,15 @@ public class DisconnectPacket: CStruct {
...
-                raw_disconnect_view.session_expiry_interval_seconds = _sessionExpiryIntervalPointer
-            }
-
+            raw_disconnect_view.session_expiry_interval_seconds = sessionExpiryIntervalPointer
+            
             return withOptionalUserPropertyArray(
                 of: userProperties) { userPropertyPointer in
 
-                if let _userPropertyPointer = userPropertyPointer {
+                if let userPropertyPointer = userPropertyPointer {
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:595]:
```suggestion
                if let userPropertyPointer {
```

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 31]
```diff
@@ -460,29 +610,30 @@ public class DisconnectPacket: CStruct {
...
-
-            let disconnectPacket = DisconnectPacket(
-                reasonCode: reasonCode,
-                sessionExpiryInterval: sessionExpiryIntervalSeconds,
-                reasonString: reasonString,
-                serverReference: serverReference,
-                userProperties: userProperties
-            )
-            return disconnectPacket
+        guard let from = from else {
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:613]:
```suggestion
        guard let from else {
```

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 32]
```diff
@@ -460,29 +610,30 @@ public class DisconnectPacket: CStruct {
...
+            return nil
         }
-        return nil
+        let disconnectView = from.pointee
+        guard let reasonCode = DisconnectReasonCode(rawValue: Int(disconnectView.reason_code.rawValue))
+        else { fatalError("aws_mqtt5_packet_disconnect_view from native missing a reason code.") }
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:618]: nit: bad format

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Client.swift [hunk 33]
```diff
@@ -65,9 +65,53 @@ public class Mqtt5Client {
...
+
     public func close() {
         self.callbackCore.close()
         aws_mqtt5_client_release(rawValue)
         rawValue = nil
     }
 }
+
+// The completion callback to invoke when an ack is received in native
+private func subscribeCompletionCallback(
```
Comment (sbSteveK) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Client.swift:102]: Does this work? This callback doesn't seem to fit the native callback signature. (packet_type, packet, error_code, context)

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 34]
```diff
@@ -460,29 +610,30 @@ public class DisconnectPacket: CStruct {
...
+
+        let sessionExpiryInterval = convertOptionalUInt32(disconnectView.session_expiry_interval_seconds)
+        let sessionExpiryIntervalSeconds: TimeInterval? = sessionExpiryInterval.map { TimeInterval($0) }
+        let reasonString = convertAwsByteCursorToOptionalString(disconnectView.reason_string)
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:622]: Don't need this function, Utilities.swift already has a helper extension. You can just do `disconnectView.reason_string.toOptionalString()`.

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 35]
```diff
@@ -578,59 +729,59 @@ public class ConnackPacket {
...
 
     static func convertFromNative(_ from: UnsafePointer<aws_mqtt5_packet_connack_view>?) -> ConnackPacket? {
+        guard let from = from else {
+            return nil
+        }
+        let connackView = from.pointee
+
+        let sessionPresent = connackView.session_present
+        guard let reasonCode = ConnectReasonCode(rawValue: Int(connackView.reason_code.rawValue))
+        else { fatalError("aws_mqtt5_packet_connack_view from native missing a reason code.") }
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:739]: bad formart, else is at the same level of the other code.

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift [hunk 36]
```diff
@@ -578,59 +729,59 @@ public class ConnackPacket {
...
+        let sessionExpiryInterval = (connackView.session_expiry_interval?.pointee).map { TimeInterval($0) }
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:740]: This can result in nil.map, does that work?

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Types.swift [hunk 37]
```diff
@@ -442,21 +449,18 @@ extension ClientOperationQueueBehaviorType {
 
 /// Optional property describing a PUBLISH payload's format.
 /// Enum values match `MQTT5 spec <https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901111>`__ encoding values.
-public enum PayloadFormatIndicator {
+public enum PayloadFormatIndicator: Int {
```
Comment (waahm7) [inline Source/AwsCommonRuntimeKit/mqtt/Mqtt5Types.swift:452]: Might want to re-discuss the API with Bret after the discussion in the Ops meeting if this is a good idea.

Comment (waahm7) [reply Source/AwsCommonRuntimeKit/mqtt/Mqtt5Types.swift:452]: If it is, we should make a struct which contains data and format for the Payload instead of having two separate connected variables.

Comment (sbSteveK) [reply Source/AwsCommonRuntimeKit/mqtt/Mqtt5Types.swift:452]: I think this enum will be extremely rarely used or set in either direction and is fully separate from payload in all other language implementations. Generally, it'll be ignored by the customer. The payload is a Data type and that feels sufficient.

Comment (waahm7) [reply Source/AwsCommonRuntimeKit/mqtt/Mqtt5Types.swift:452]: Do we even need this indicator? We should just pass the bytes as it is.

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift (hunk 28 above)
Comment (xiazhvera) [reply Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:439]: As userProperties is a class member, I dont think it worked, while I tried with following:
```
if let userPropertyPointer,
   let userPropertyCount = self.userProperties?.count {

}
```

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift (hunk 34 above)
Comment (xiazhvera) [reply Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:622]: This change will be addressed in a separate PR.

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Types.swift (hunk 37 above)
Comment (sbSteveK) [reply Source/AwsCommonRuntimeKit/mqtt/Mqtt5Types.swift:452]: I don't think we need it since like you said, it'll always be bytes that can be converted to utf8 strings... but it's also part of the mqtt5 spec and expected to be there. It's possible someone may want to check its value for... something?

File: Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift (hunk 36 above)
Comment (xiazhvera) [reply Source/AwsCommonRuntimeKit/mqtt/Mqtt5Packets.swift:745]: `map` works for Optional: https://developer.apple.com/documentation/swift/optional/map(_:)
